### PR TITLE
Add image to manifests and CLIs

### PR DESCRIPTION
**Pull request by sagerb**
# Description

These commands now accept the `--image` or `-m` parameter, to specify the content image
- `rsconnect write-manifest api`: 
- `rsconnect write-manifest bokeh`: 
- `rsconnect write-manifest dash`:
- `rsconnect write-manifest fastapi`:
- `rsconnect write-manifest notebook`:
- `rsconnect write-manifest quarto`:
- `rsconnect write-manifest streamlit`:

- `rsconnect deploy api`
- `rsconnect deploy bokeh`
- `rsconnect deploy dash`
- `rsconnect deploy fastapi` 
- `rsconnect deploy notebook` 
- `rsconnect deploy quarto`  
- `rsconnect deploy streamlit`  

Not impacted (but related) and unaffected (do not accept the --image parameter)
- `rsconnect deploy manifest`
- `rsconnect deploy other-content`
- `rsconnect deploy html`

Fixes rstudio/connect#21381

# Implementation Notes

The objective of this PR was to add a single command line option to multiple commands. Unfortunately, the refactoring involved encountered a number of early errors which highlighted the need to enable and utilize better development tooling as part of the refactoring process. After discussions with multiple folks who were focused on the rsconnect-python codebase, we made the decision to gain confidence by making two major changes in the areas where the code was being impacted:
- Use of 3.5 compatible type hints - basically adding type definitions to the impacted functions (and trying to stay away from the use of `Any`
- Removal of optional / default parameters from the function declarations, so that tooling would be able to identify missing parameters. The actual use of optional parameters (often time un-named) was limited to test code, so the refactoring involved in this was rather small for the benefit we gained.

In addition, these changes allowed us to turn on linting for our test code, which was also calling these refactored functions. 

After these changes were made, our confidence of being able to identify issues with the refactoring `statically` rather than `functionality` was greatly improved. Reproduction of the manually encountered errors within the new codebase were successfully identified by the `make lint` build step and did not require actual execution of the code path having the problem.

The changes, therefore, within this PR are a bit larger than one would think for simply adding an option or two to the commands listed above. They were, however, critical to improving our confidence in making this change (where in the past, this was a recognized fragile effort).

I have also added notes into the `CONTRIBUTE.md` file which should help bootstrap development efforts for this repo. These were previously missing and did represent an initial barrier to contribution.

# Testing Notes / Validation Steps

This feature does not make any changes to the APIs being used, so it should not have introduced any new compatibility requirements w/ a Connect version. The new options, present on the CLI, will happily be accepted by rsconnect-python regardless how the Connect server is configured. The impact, however, of how they are used by the server, differs by the Connect release and configuration. Starting with the May 2022 release, the Connect server when configured with `Launcher.Enabled = true` and `Launcher.Kubernetes = true`, will attempt to use the `environment.image` setting within the content bundle to match the target content image. If this configuration is not active or if it is a previous version of Connect, that field will be ignored and the legacy behavior will not be impacted. Because of this.. **it is prudent to run a few of these commands against an older version or a non-kubernetes configuration, just to make sure there is no compatibility issue**. Those steps are NOT outlined below.

## Approach taken
By using a kubernetes configuration for Connect, which utilizes two duplicate images, we are able to verify the functionality of which image is used by the server when the image flag is specified and when it is not.

Verification for each of the impacted commands is grouped by target, allowing us to focus on a target at a time, but executing fairly similar steps for each target:

### Validate write-manifest <target>
- Validate the help output contains the image option
- Update the reference manifest.json to be reflective of the active version of python, save this off.
- Use `rsconnect write-manifest -image rstudio/dev-connect-duplicate <target>` to update the manifest.json to include the `environment.image` setting
- Diff the manifest.json files (with and without image) to confirm there are no other significant differences
- 

### Validate deploy manifest for the manifest.json with the image setting
- Use `rsconnect deploy manifest <manifest file>` to confirm proper deployment

### Validate deploy <target> with and without image
- Validate the help output contains image
- Use `rsconnect deploy <target>` to validate that target is able to be deployed to server and that the server selects the content image target by using the built-in algorithm (will select `rstudio/dev-connect`).
- Use `rsconnect deploy <target> --image rstudio/dev-connect-duplicate` to validate that target is able to be deployed to server and that the server uses the image option to determine the target image (will be `rstudio/dev-connect-duplicate`).

## Testing details

Need to be setup specifically with Python 3.8.x to work with the test data / content outlined below.

```bash
brew install python@3.8
#Updated my path to have `/usr/local/opt/python@3.8/bin` ahead of `/usr/local/bin` in ~/.zshenv 
source ~/.zshenv
```

Setup virtual environment to use branch rsconnect within the base directory of rsconnect-python repo:

```bash
# Setup a virtual python environment w/ 3.8.x
rm -rd .venv
python3 --version        
# confirm version listed above is 3.8.x
python3 -m venv .venv
# Activate the virtual environment
source .venv/bin/activate
python3 --version        
# confirm version listed above is 3.8.x
# install our requirements into the virtual environment
pip install -r requirements.txt
# install rsconnect-python with a symbolic link to the locations repository, 
# meaning any changes to code in there will automatically be reflected
pip install -e ./
# clear out path to executables
hash -r
```

For Connect, I recommend using the new `dev-connect` docker file workflow, which will be available from `main` before this PR goes to testing, as will David's supporting server branch which implements the image logic needed on the server.

To start the server, you'll need to:

1. Build the codebase, from the root of the rstudio/connect project:
    - `just ui/dashboard/clean`
    - `just bootstrap build`
2. Update `yeti/docker/built-in/connect-shared-files/rsc-launcher-runtime.yml` to have the following two additional image specifications, to be placed at top of image list (on line 3, after the line `images:`). This should move all of the other image entries down in the file / list.

```
  - 
    name: rstudio/dev-connect
    python:
      installations:
        - path: /opt/python/3.8.12/bin/python3.8
          version: 3.8.12
    r:
      installations:
        - path: /opt/R/3.6.3/bin/R
          version: 3.6.3
    quarto:
      installations:
        - path: /opt/rstudio-connect/ext/quarto/bin/quarto
          version: 0.9.16
  - 
    name: rstudio/dev-connect-duplicate
    python:
      installations:
        - path: /opt/python/3.8.12/bin/python3.8
          version: 3.8.12
    r:
      installations:
        - path: /opt/R/3.6.3/bin/R
          version: 3.6.3
    quarto:
      installations:
        - path: /opt/rstudio-connect/ext/quarto/bin/quarto
          version: 0.9.16
  -
```

3. Create a custom template and add an `imagePullPolicy=Never` to the pods volume definition.
    - Copy `packaging/launcher/connect-k8s-templates/job.tpl` to `config/launcher/k8s-custom-templates`
    - Edit `config/launcher/k8s-custom-templates/job.tpl` and add the following line after line 83 (`image: {{ toYaml .Job.container.image }}`) - should be indented same level as the image line.

```
          imagePullPolicy: Never
```

    - Save the file with your changes.
4. Temporarily update the main justfile to have (updating area around 228 of `justfile`):

```
    # Always build the docker images
    docker build . -t rstudio/dev-connect -f docker/devconnect/Dockerfile
    if [[ "${USE_KUBERNETES}" == "yes" ]]; then
        docker build . -t rstudio/dev-connect-duplicate -f docker/devconnect/Dockerfile
    fi
```

5. Reset the dev-connect data volume:
    - `docker volume rm dev-connect-volume`
6. Reset your kubernetes cluster:
    - `kubectl -n rsc-dev delete jobs --all`
    - `kubectl -n rsc-dev delete svc --all`
7. Start the server:
    - `USE_KUBERNETES=yes just start-dev-connect`

### Verify write-manifest api

WITH the venv active.. cd into your rstudio/connect-content repo.
From:  rstudio/connect-content
1. cd bundles/python-flaskapi
2. execute:

```bash
rsconnect write-manifest api --help
```

and confirm that `--image` and `-m` is listed in the help

3. execute: 

```bash
cd bundles/python-flaskapi
rsconnect write-manifest api --help
```

Confirm that `--image` and `-m` is listed in the help

4. execute:

```bash
# baseline the manifest w/ your virtual environment basics
rsconnect write-manifest api . --overwrite
mv manifest.json manifest-original.json
rsconnect write-manifest api --image rstudio/dev-connect-duplicate .
diff manifest.json manifest-original.json
```

confirm that you see the environment/image path added in the new manifest. This will look something like:

```
16,18d15
<   "environment": {
<     "image": "rstudio/dev-connect-duplicate"
<   },
26,28d22
<     "manifest-original.json": {
<       "checksum": "f5dcb185f388ab9694ac4e6e3ce9f8e0"
<     },
```

where the first section is the important verification (and the second is just an artifact that we added a file).

### Verify deploy manifest

```bash
rsconnect deploy manifest -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new ./manifest.json
```

Verify output is similar to:

```
Deployment log:
Building Python API...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify deploy api

WITH your terminal working directory same as previously (bundles/python-flaskapi)

1. Confirm that the help output contains the image option:

```bash
rsconnect deploy api --help
```

2. execute:

```bash
rsconnect deploy api -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new .
```

Confirm that the automatic image was selected, by verifying the output is similar to:

```
Deployment log:
Building Python API...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect which has version 3.8.8
```

3. execute:

```bash
rsconnect deploy api -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --image rstudio/dev-connect-duplicate --new .
```

Verify the output is similar to the following:

```
Deployment log:
Building Python API...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify write-manifest bokeh

WITH the venv active.. cd into your rstudio/connect-content repo.
From:  rstudio/connect-content:

```bash
cd bundles/python-bokeh
rsconnect write-manifest bokeh --help
# confirm that `--image` and `-m` is listed in the help
# baseline the manifest w/ your virtual environment basics
rsconnect write-manifest bokeh . --overwrite
mv manifest.json manifest-original.json
rsconnect write-manifest bokeh --image rstudio/dev-connect-duplicate .
diff manifest.json manifest-original.json
```

Confirm that you see the environment/image path added in the new manifest. This will look something like the following

```
16,18d15
<   "environment": {
<     "image": "rstudio/dev-connect-duplicate"
<   },
26,28d22
<     "manifest-original.json": {
<       "checksum": "f5dcb185f388ab9694ac4e6e3ce9f8e0"
<     },
```
Where the first section is the important verification (and the second is just an artifact that we added a file).

### Verify deploy manifest

```bash
rsconnect deploy manifest -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new  ./manifest.json
```

Verify output is similar to:

```
Deployment log:
Building Bokeh application...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify deploy bokeh

WITH your terminal working directory same as previously (bundles/python-bokeh)
1. Confirm that the help output contains the image option:

```bash
rsconnect deploy bokeh --help
```

2. execute:

```bash
rsconnect deploy bokeh -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new .
```

Confirm that the automatic image was selected, by verifying the output is similar to:

```
Deployment log:
Building Bokeh application...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect which has version 3.8.8
```

3. execute:

```bash
rsconnect deploy bokeh -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --image rstudio/dev-connect-duplicate --new .
```

Verify the output is similar to the following:

```
Deployment log:
Building Bokeh application...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify write-manifest dash

WITH the venv active.. cd into your rstudio/connect-content repo.
From:  rstudio/connect-content:

1. Execute:

```bash
cd bundles/python-dash
rsconnect write-manifest dash --help
```

Confirm that `--image` and `-m` is listed in the help

2. Execute:

```bash
# baseline the manifest w/ your virtual environment basics
rsconnect write-manifest dash . --overwrite
mv manifest.json manifest-original.json
rsconnect write-manifest dash --image rstudio/dev-connect-duplicate .
diff manifest.json manifest-original.json
```

Confirm that you see the environment/image path added in the new manifest. This will look something like the following
```
16,18d15
<   "environment": {
<     "image": "rstudio/dev-connect-duplicate"
<   },
26,28d22
<     "manifest-original.json": {
<       "checksum": "f5dcb185f388ab9694ac4e6e3ce9f8e0"
<     },
```
Where the first section is the important verification (and the second is just an artifact that we added a file).

### Verify deploy manifest

```bash
rsconnect deploy manifest -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new  ./manifest.json
```

Verify output is similar to:

```
Deployment log:
Building Dash application...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify deploy dash

WITH your terminal working directory same as previously (bundles/python-dash)

1. Confirm that the help output contains the image option:

```bash
rsconnect deploy dash --help
```

2. execute:

```bash
rsconnect deploy dash -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new .
```

Confirm that the automatic image was selected, by verifying the output is similar to:

```
Deployment log:
Building Dash application...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect which has version 3.8.8
```

3. execute:

```bash
rsconnect deploy dash -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --image rstudio/dev-connect-duplicate --new .
```

Verify the output is similar to the following:

```
Deployment log:
Building Dash application...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify write-manifest fastapi

WITH the venv active.. cd into your rstudio/connect-content repo.
From:  rstudio/connect-content:

1. Execute:
 
```bash
cd bundles/fastapi-simple
rsconnect write-manifest fastapi --help
```

Confirm that `--image` and `-m` is listed in the help

2. Execute:

```bash
# baseline the manifest w/ your virtual environment basics
rsconnect write-manifest fastapi . --overwrite
mv manifest.json manifest-original.json
rsconnect write-manifest fastapi --image rstudio/dev-connect-duplicate .
diff manifest.json manifest-original.json
```

Confirm that you see the environment/image path added in the new manifest. This will look something like the following

```
16,18d15
<   "environment": {
<     "image": "rstudio/dev-connect-duplicate"
<   },
26,28d22
<     "manifest-original.json": {
<       "checksum": "f5dcb185f388ab9694ac4e6e3ce9f8e0"
<     },
```

Where the first section is the important verification (and the second is just an artifact that we added a file).

### Verify deploy manifest

```bash
rsconnect deploy manifest -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new  ./manifest.json
```

Verify output is similar to:

```
Deployment log:
Building FastAPI application...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify deploy fastapi

WITH your terminal working directory same as previously (bundles/fastapi-simple)

1. Confirm that the help output contains the image option:

```bash
rsconnect deploy fastapi --help
```

2. execute:

```bash
rsconnect deploy fastapi -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new .
```

Confirm that the automatic image was selected, by verifying the output is similar to:

```
Deployment log:
Building FastAPI application...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect which has version 3.8.8
```

3. execute:

```bash
rsconnect deploy fastapi -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --image rstudio/dev-connect-duplicate --new .
```

Verify the output is similar to the following:

```
Deployment log:
Building FastAPI application...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify write-manifest notebook

WITH the venv active.. cd into your rstudio/connect-content repo.
From:  rstudio/connect-content:

1. Execute:

```bash
cd bundles/stock-report-jupyter
rsconnect write-manifest notebook --help
```

Confirm that `--image` and `-m` is listed in the help

2. Execute:

```bash
# baseline the manifest w/ your virtual environment basics
rsconnect write-manifest notebook --overwrite stock-report-jupyter.ipynb quandl-wiki-tsla.json.gz thumbnail.jpg
mv manifest.json manifest-original.json
rsconnect write-manifest notebook --image rstudio/dev-connect-duplicate stock-report-jupyter.ipynb quandl-wiki-tsla.json.gz thumbnail.jpg
diff manifest.json manifest-original.json
```

Confirm that you see the environment/image path added in the new manifest. This will look something like the following

```
16,18d15
<   "environment": {
<     "image": "rstudio/dev-connect-duplicate"
<   },
```
Where the first section is the important verification (and the second is just an artifact that we added a file).

### Verify deploy manifest

```bash
rsconnect deploy manifest -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new  ./manifest.json
```

Verify output is similar to:

```
Deployment log:
Building Jupyter notebook...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify deploy notebook

WITH your terminal working directory same as previously (bundles/stock-report-jupyter)

1. Confirm that the help output contains the image option:

```bash
rsconnect deploy notebook --help
```

2. execute:

```bash
rsconnect deploy notebook -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new stock-report-jupyter.ipynb quandl-wiki-tsla.json.gz thumbnail.jpg
```

Confirm that the automatic image was selected, by verifying the output is similar to:

```
Deployment log:
Building Jupyter notebook...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect which has version 3.8.8
```

3. execute:

```bash
rsconnect deploy notebook -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --image rstudio/dev-connect-duplicate --new stock-report-jupyter.ipynb quandl-wiki-tsla.json.gz thumbnail.jpg
```

Verify the output is similar to the following:

```
Deployment log:
Building Jupyter notebook...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify write-manifest quarto

WITH the venv active.. cd into your rstudio/connect-content repo.
From:  rstudio/connect-content:

1. Execute:

```bash
cd bundles/quarto-proj-py
rsconnect write-manifest quarto --help
```

Confirm that `--image` and `-m` is listed in the help

2. Execute:

```bash
# baseline the manifest w/ your virtual environment basics
rsconnect write-manifest quarto . --overwrite
mv manifest.json manifest-original.json
rsconnect write-manifest quarto --image rstudio/dev-connect-duplicate .
diff manifest.json manifest-original.json
```

Confirm that you see the environment/image path added in the new manifest. This will look something like the following

```
16,18d15
<   "environment": {
<     "image": "rstudio/dev-connect-duplicate"
<   },
26,28d22
<     "manifest-original.json": {
<       "checksum": "f5dcb185f388ab9694ac4e6e3ce9f8e0"
<     },
```
Where the first section is the important verification (and the second is just an artifact that we added a file).

### Verify deploy manifest

```bash
rsconnect deploy manifest -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new  ./manifest.json
```

Verify output is similar to:

```
Deployment log:
Building Quarto document...
Bundle created with Python version 3.8.13 and Quarto version 0.9.16 is compatible with environment Kubernetes::rstudio/dev-connect with Python version 3.8.12 from /opt/python/3.8.12/bin/python3.8 and Quarto version 0.9.16 from /opt/rstudio-connect/ext/quarto/bin/quarto 
Bundle requested Python version 3.8.13; using /opt/python/3.8.12/bin/python3.8 from Kubernetes::rstudio/dev-connect which has version 3.8.12
```

### Verify deploy quarto

WITH your terminal working directory same as previously (bundles/quarto-project-py)

1. Confirm that the help output contains the image option:

```bash
rsconnect deploy quarto --help
```

2. execute:

```bash
rsconnect deploy quarto -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new .
```

Confirm that the automatic image was selected, by verifying the output is similar to:

```
Deployment log:
Building Quarto document...
Bundle created with Python version 3.8.13 and Quarto version 0.9.16 is compatible with environment Kubernetes::rstudio/dev-connect with Python version 3.8.12 from /opt/python/3.8.12/bin/python3.8 and Quarto version 0.9.16 from /opt/rstudio-connect/ext/quarto/bin/quarto 
Bundle requested Python version 3.8.13; using /opt/python/3.8.12/bin/python3.8 from Kubernetes::rstudio/dev-connect which has version 3.8.12
```

3. execute:

```bash
rsconnect deploy quarto -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --image rstudio/dev-connect-duplicate --new .
```

Verify the output is similar to the following:

```
Deployment log:
Building Quarto document...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 and Quarto version 0.9.16 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.12 from /opt/python/3.8.12/bin/python3.8 and Quarto version 0.9.16 from /opt/rstudio-connect/ext/quarto/bin/quarto 
Bundle requested Python version 3.8.13; using /opt/python/3.8.12/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.12
```

### Verify write-manifest streamlit

WITH the venv active.. cd into your rstudio/connect-content repo.
From:  rstudio/connect-content:

1. Execute:

```bash
cd bundles/python-streamlit
rsconnect write-manifest streamlit --help
```

Confirm that `--image` and `-m` is listed in the help

```bash
# baseline the manifest w/ your virtual environment basics
rsconnect write-manifest streamlit --overwrite .
mv manifest.json manifest-original.json
rsconnect write-manifest streamlit --image rstudio/dev-connect-duplicate .
diff manifest.json manifest-original.json
```

Confirm that you see the environment/image path added in the new manifest. This will look something like the following

```
16,18d15
<   "environment": {
<     "image": "rstudio/dev-connect-duplicate"
<   },
```
Where the first section is the important verification (and the second is just an artifact that we added a file).

### Verify deploy manifest

```bash
rsconnect deploy manifest -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new  ./manifest.json
```

Verify output is similar to:

```
Deployment log:
Building Streamlit application...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### Verify deploy streamlit

WITH your terminal working directory same as previously (bundles/python-streamlit)

1. Confirm that the help output contains the image option:

```bash
rsconnect deploy streamlit --help
```

2. execute:

```bash
rsconnect deploy streamlit -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --new .
```

Confirm that the automatic image was selected, by verifying the output is similar to:

```
Deployment log:
Building Streamlit application...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect with Python version 3.8.12 from /opt/python/3.8.12/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.12/bin/python3.8 from Kubernetes::rstudio/dev-connect which has version 3.8.12
```

3. execute:

```bash
rsconnect deploy streamlit -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --image rstudio/dev-connect-duplicate --new .
```

Verify the output is similar to the following:

```
Deployment log:
Building Streamlit application...
A target image was specified: rstudio/dev-connect-duplicate, checking if it is compatible...
Bundle created with Python version 3.8.13 is compatible with environment Kubernetes::rstudio/dev-connect-duplicate with Python version 3.8.8 from /opt/python/3.8.8/bin/python3.8 
Bundle requested Python version 3.8.13; using /opt/python/3.8.8/bin/python3.8 from Kubernetes::rstudio/dev-connect-duplicate which has version 3.8.8
```

### deploy manifest 

from the bundles/python-streamlit subdirectory

```bash
rsconnect deploy manifest -s http://localhost:3939 -k c7achTdggWVGtr851azThcyYDwH5JRgf --image rstudio/dev-connect-duplicate ./manifest.json
```

Confirm you see an error, similar to:

```
Usage: rsconnect deploy manifest [OPTIONS] FILE
Try 'rsconnect deploy manifest --help' for help.

Error: No such option: --image
```

Confirm help doesn't include image

```bash
rsconnect deploy manifest --help
```

It should not include a --image option

### deploy other-content

from the bundles/python-streamlit subdirectory

```bash
rsconnect deploy other-content --image rstudio/dev-connect-duplicate ./manifest.json
```

Confirm you see an error, similar to:

```
Usage: rsconnect deploy manifest [OPTIONS] FILE
Try 'rsconnect deploy manifest --help' for help.

Error: No such option: --image
```

Confirm help doesn't include image

```bash
rsconnect deploy other-content --help
```

It should not include a --image option

### CODE REVIEW

File: tests/test_metadata.py [hunk 1]
```diff
@@ -10,9 +11,19 @@
...
+        self.server_store_path = join(temp, "servers.json")
+        self.assertFalse(exists(self.server_store_path))
+
         self.server_store.set("foo", "http://connect.local", "notReallyAnApiKey", ca_data="/certs/connect")
         self.server_store.set("bar", "http://connect.remote", "differentApiKey", insecure=True)
+        self.assertEqual(len(self.server_store.get_all_servers()), 2, "Unexpected servers after setup")
+
+    def tearDown(self):
+        # clean up our temporary store
+        remove(self.server_store_path)
```
Comment (mmarchetti) [inline tests/test_metadata.py:26]: `os.remove` doesn't handle directories.  `shutil.rmtree` will handle non-empty directories.

Comment (sagerb) [reply tests/test_metadata.py]: Thanks! Fixed!

**Comment by mmarchetti** [comment]
We've removed the API docs for now since they suggested that every function in the package was part of the API - it's currently still up on docs.rstudio.com because we haven't released since then. We still need to create a new "how to use the Python API" type of doc.

File: tests/target_image/test.sh [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	# 	else
+	# 		echo "nope"
+	# 	fi
+	# done
+}
+
+ConfirmManifestDiff
+
+
+# ConfirmHelp
```
[the remark this answers is not in the document]
Comment (sagerb) [reply tests/target_image/test.sh:80]: absolutely! Thanks!

File: tests/target_image/test.sh [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	fi
+	return 1
+}
+
+function ConfirmHelp() {
+	declare -a StringArray=(
+		"rsconnect write-manifest streamlit --help"
+		"rsconnect write-manifest fastapi --help" 
+		"rsconnect deploy streamlit --help"
+		"rsconnect deploy fastapi --help"
```
[the remark this answers is not in the document]
Comment (sagerb) [reply tests/target_image/test.sh]: ended up removing this script.. it wasn't complete and was only a start. I'm going to keep what I created locally, but I'll be working with Kevin to get integration tests which utilize a kubernetes launcher configuration.

**Pull request merged**